src/esi/Expression.cc

```diff
@@ -97,6 +100,16 @@ stackpop(stackmember * s, int *depth)
     cleanmember(&s[*depth]);
 }
 
+/// push item onto the stack.
+/// Throws an exception with the given message if depth is not referencing
+/// a valid location of the stack both before and after the push.
+static void
+stackpush(stackmember *stack, stackmember &item, int *depth, const char *message)
+{
+    Must2(*depth >= 0 && *depth < STACK_DEPTH_LIMIT-1, message);
```

Comment:
This check seems too conservative because the `depth` parameter can reach `STACK_DEPTH_LIMIT`, right?

Also the first check does not seem to be related to the `message` being passed to this method.

```suggestion
    assert(*depth >= 0);
    Must2(*depth < STACK_DEPTH_LIMIT, message);
```

Reply:
Remember that the key issue behind this change is that the depth can be remotely controlled. Both push and pop can be manipulated by the server response.

Reply:
BY that aI meant that we can split this into two MUST, but using assert is not an option.

Reply:
> we can split this into two

Yes, and we should. Please do that.

> using assert is not an option

Whether the first condition should be an assert() or Must() depends on whether Squid has already checked/ensured that depth does not go negative. If we are confident that Squid has, then this check can and should be an assert(). Otherwise, this check should be a Must(). Assertions are overall better (where they meet their usage preconditions), but Must()s are virtually always acceptable, so if you are not sure about `depth` checks in callers, then you can use a Must() instead of an assert here.

Reply:
resolved with the the exception model